### ISSUE TITLE
Training is not moving ahead after First Batch

### ISSUE THREAD

**Issue by bhadreshpsavani** [pytorch/xla#2439]
## ❓ Questions and Help
When I train Transformer T5 model on Kaggle TPU, It is not moving to the next batch after the first batch,

This issue all persist when I train it on Colab.

Note: MXU status is also always showing Zero
```
[xla:0], Loss=9.392
Epoch: 0, Loss:  9.392439842224121
[xla:2], Loss=10.457
[xla:7], Loss=9.749
[xla:4], Loss=9.688
[xla:6], Loss=10.734
[xla:3], Loss=8.620
[xla:5], Loss=9.350
[xla:1], Loss=9.519
```
Here is hardware usage status:

![image](https://user-images.githubusercontent.com/26653468/90480069-58fe1780-e14d-11ea-8f79-a27890c9cd31.png)

[Notebook
](https://www.kaggle.com/bhadreshsavani/naturalqat5pytorchtpu/execution)

**Comment by JackCaoG** [comment]
Can you share a colab so we can repo the issue?

**Comment by bhadreshpsavani** [comment]
Hi @JackCaoG ,
Here is the [colab notebook](https://github.com/bhadreshpsavani/EfficientQAExperiments/blob/master/natural_qa_t5_pytorch_tpu.ipynb)

**Comment by cjolivier01** [comment]
I don't think you need to do the .to_device for each batch of data -- ParallelLoader should be delivering it to you already assigned to the device.  Did you have problems with it before?

I believe your freezing issue may be due to the xm.master_print not being inside of a step closure, so this means that only on the master it will compile/switch graphs and any rendezvous that occurs between the multiple devices will block forever since only the master has changed graph and thus tried to rendezvous.

Of course, I am not as knowledgeable as @JackCaoG so I may be wrong about this, but this is what it looks like to me is going on.

**Comment by JackCaoG** [comment]
Thanks @cjolivier01 for the quick and detail answer. @bhadreshpsavani ParallelLoader will take cpu tensor and then send it to the device. You should not need to perform any `to(device)` on the data returned by the `ParallelLoader`(and it might cause problem). Please take a look at https://github.com/pytorch/xla/blob/master/contrib/colab/mnist-training.ipynb for the example usage of loader.

@cjolivier01 's point about master print is right. Printing loss will result in early evaluation of the loss in master process and caused the unbalance graph. It is recommended to do it after the device data being materialized (in a [step closure](https://github.com/pytorch/xla/blob/0c34871b91159e51603119c42d9fe58fb7d1da88/torch_xla/core/xla_model.py#L674) ). You can also try plain print, which will force all processes evaluate the loss early.

**Comment by cjolivier01** [comment]
I was thinking about this, and it's probably the case that the f-string is being evaluated by all workers even though only printed on master.  If you do a print('step') at the top of each training step, it can be seen that it is moving very slowly.  I didn't check thoroughly in your model code, but it's possible that you are compiling and/or switching graphs multiple times per step, causing it to train very slowly.  For example, if item() is being called on any variables or one_hot() is used :)
Looks like it is moving very slowly:
step: 0, device: xla:0
step: 0, device: xla:0
step: 0, device: xla:0
step: 1, device: xla:1
step: 1, device: xla:0
step: 1, device: xla:0
step: 1, device: xla:0
step: 1, device: xla:0
step: 1, device: xla:0
step: 1, device: xla:0
step: 1, device: xla:0
step: 2, device: xla:0
step: 2, device: xla:0
step: 2, device: xla:1
step: 2, device: xla:0
step: 2, device: xla:0
step: 2, device: xla:0
step: 2, device: xla:0
step: 2, device: xla:0
step: 3, device: xla:0
step: 3, device: xla:0
step: 3, device: xla:0
step: 3, device: xla:0
step: 3, device: xla:1
step: 3, device: xla:0
step: 3, device: xla:0
step: 3, device: xla:0

**Comment by bhadreshpsavani** [comment]
Hi @cjolivier01,
I made a few changes and realize that it's not stopped, it's moving really slow. I am not using `item()` or `one_hot()` although but my training loop has `contiguous()` and `clone().detach()` that might cause this issue, I am not sure.

**Comment by cjolivier01** [comment]
@bhadreshpsavani thanks for the update. I had printed out steps and also tried on one TPU and also noticed it was moving very slowly. In timing the training loop pass it didn’t look especially like it was compiling more than once, although this could be further verified by getting the compile/execute stats from the metrics module.

Does it run that slowly on CPU and/or GPU? I didn’t dig into the actual model as it looks to be in another module.
